File: Algorithms/1004_Max_Consecutive_Ones_III/Python/Max_Consecutive_Ones_III_Solution_1.py

# Space: O(1)
# Time: O(n)

class Solution:
    def longestOnes(self, nums, k):
        length = len(nums)
        if length == 0: return 0
        if length == 1: return 1 if k >= 1 else 0
        fast, slow, res, count = 0, 0, 0, k
        while fast < length:
            if nums[fast] == 1:
                fast += 1
            elif nums[fast] == 0 and count > 0:
                count -= 1
                fast += 1

            elif count == 0:
                res = max(len(nums[slow:fast]), res)

                while count == 0:
                    if nums[slow] == 0:
                        count += 1
                    slow += 1
        return res

File: Algorithms/1004_Max_Consecutive_Ones_III/Java/MaxConsecutiveOnesIIISolution1.java

// Space: O(1)
// Time: O(n)

class Solution {
    public int longestOnes(int[] nums, int k) {
        if (nums.length == 0) return 0;
        if (nums.length == 1) return k > 0 ? 1 : 0;
        int slow = 0, fast = 0, res = 0, count = k;
        while (fast < nums.length) {
            if (nums[fast] == 0 && count > 0) {
                count--;
                fast++;
            } else if (nums[fast] == 1) {
                fast++;
            } else if (count == 0) {
                res = Math.max(fast - slow, res);
                if (nums[slow] == 0) count++;
                slow++;
            }
        }
        res = Math.max(fast - slow, res);
        return res;
    }
}




